simulator/controller/controller.go

```diff
@@ -0,0 +1,173 @@
+	ControllerStartJitter = 1.0
+)
+
+type InitFunc func(controllerCtx ControllerContext) error
+
+// ControllerInitializersFunc is used to create a collection of initializers
+//  given the loopMode.
+//nolint:revive // Intentionally used the same name.
+type ControllerInitializersFunc func() (initializers map[string]InitFunc)
+
+var _ ControllerInitializersFunc = NewControllerInitializers
+
+func NewController(client clientset.Interface, c *restclient.Config) (func(), error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	err := mustSetupScheduler(client, ctx.Done())
+	if err != nil {
+		cancel()
+		return nil, xerrors.Errorf("call mustSetupScheduler func: %w", err)
+	}
+
+	go Run(client, c, ctx.Done())
+	shutdownFunc := func() {
+		cancel()
+	}
+
+	return shutdownFunc, nil
+}
+
+// mustSetupScheduler starts a controller which is required to run scheduler.
+func mustSetupScheduler(client clientset.Interface, stopCh <-chan struct{}) error {
+	err := StartPersistentVolumeController(client, stopCh)
+	if err != nil {
+		return xerrors.Errorf("start pv controller: %w", err)
+	}
+	return nil
+}
+
+// Run runs the KubeControllerManagerOptions.  This should never exit.
+func Run(client clientset.Interface, config *restclient.Config, stopCh <-chan struct{}) {
+	run := func(ctx context.Context, initializersFunc ControllerInitializersFunc) {
+		controllerContext, err := CreateControllerContext(client, config, ctx.Done())
+		if err != nil {
+			klog.Fatalf("error building controller context: %v", err)
+		}
+		controllerInitializers := initializersFunc()
+		if err := StartControllers(controllerContext, controllerInitializers); err != nil {
+			klog.Fatalf("error starting controllers: %v", err)
+		}
+
+		controllerContext.InformerFactory.Start(stopCh)
+		controllerContext.ObjectOrMetadataInformerFactory.Start(stopCh)
+		close(controllerContext.InformersStarted)
+
+		select {}
+	}
+	run(context.Background(), NewControllerInitializers)
+	panic("unreachable")
+}
+
+// StartControllers starts a set of controllers with a specified ControllerContext.
+func StartControllers(ctx ControllerContext, controllers map[string]InitFunc) error {
+	for controllerName, initFn := range controllers {
+		time.Sleep(wait.Jitter(ctx.ComponentConfig.Generic.ControllerStartInterval.Duration, ControllerStartJitter))
+
+		klog.Infof("Starting %q", controllerName)
+		err := initFn(ctx)
+		if err != nil {
+			klog.Errorf("Error starting %q", controllerName)
+			return xerrors.Errorf("starting %v: %w", controllerName, err)
+		}
+		klog.Infof("Started %q", controllerName)
+	}
+	return nil
+}
+
+// NewControllerInitializers is a public map of named controller groups paired to their InitFunc.
+// This allows for structured downstream composition and subdivision.
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["deployment"] = StartDeploymentController
+	controllers["replicaset"] = StartReplicaSetController
```

Comment:
I think pvcontroller also can be started with this controller initializes.
There seems to be no reason to separate the logic to start controller for pv controller and deployment/replicaset controller.